File: problems/04/best_time_to_buy_and_sell_stock.c
#include <criterion/criterion.h>

int buy_sell(int *nums, int len)
{
  int profit = 0;

  for (int left = 0, right = 1; left < len && right < len; right++)
  {
    int current = nums[right] - nums[left];
    if (current > profit)
    {
      profit = current;
    }
    else if (current < 0)
    {
      left = right;
    }
  }

  return profit;
}

Test(BuySell, BuySellTests)
{
  int t1[6] = {7, 1, 5, 3, 6, 4};
  cr_expect(5 == buy_sell(t1, 6));

  int t2[5] = {7, 6, 4, 3, 1};
  cr_expect(0 == buy_sell(t2, 5));

  cr_expect(0 == buy_sell(NULL, 0));

  int t3[1] = {1};
  cr_expect(0 == buy_sell(t3, 1));

  int t4[2] = {1, 0};
  cr_expect(0 == buy_sell(t4, 2));
}

File: problems/04/best_time_to_buy_and_sell_stock.exs
defmodule BuySellStock do
  def solve(nums) do
    max_profit(nums, 0, nil)
  end

  defp max_profit([], max_profit, _min_price), do: max_profit

  defp max_profit([price | rest], max_profit, nil) do
    max_profit(rest, max_profit, price)
  end

  defp max_profit([price | rest], max_profit, min_price) do
    profit = price - min_price
    max_profit = max(max_profit, profit)
    min_price = min(min_price, price)
    max_profit(rest, max_profit, min_price)
  end
end

defmodule Tests do
  use ExUnit.Case

  test "something" do
    assert BuySellStock.solve([7, 1, 5, 3, 6, 4]) == 5
    assert BuySellStock.solve([7, 6, 4, 3, 1]) == 0
    assert BuySellStock.solve([]) == 0
  end
end
